src/main/python/mmlspark/cyber/feature/indexers.py

```diff
@@ -0,0 +1,143 @@
+            output_col,
+            f.when(f.col(output_col).isNotNull(), f.col(output_col)).otherwise(f.lit(0))
+        ).drop(
+            input_col
+        )
+
+
+class IdIndexer(Estimator, HasInputCol, HasOutputCol):
+    partitionKey = Param(
+        Params._dummy(),
+        "partitionKey",
+        "The name of the column to partition by, i.e., make sure the indexing takes the partition into account. "
+        "This is exemplified in reset_per_partition."
+    )
+
+    resetPerPartition = Param(
+        Params._dummy(),
+        "resetPerPartition",
+        "When set to True then indexing is consecutive from [1..n] for each value of the partition column. "
+        "When set to False then indexing is consecutive for all partition and column values."
+    )
+
+    def __init__(self, input_col: str, partition_key: str, output_col: str, reset_per_partition: bool):
+        super().__init__()
+
+        ExplainBuilder.build(
+            self,
+            inputCol=input_col,
+            partitionKey=partition_key,
+            outputCol=output_col,
+            resetPerPartition=reset_per_partition
+        )
+
+    def _make_vocab_df(self, df):
+        ucols = [self.getPartitionKey(), self.getInputCol()]
+        the_df = df.select(ucols).distinct().orderBy(ucols)
+
+        return DataFrameUtils.zip_with_index(
+            df=the_df,
+            start_index=1,
+            col_name=self.getOutputCol(),
+            partition_col=self.getPartitionKey(),
+            order_by_col=self.getInputCol()
+        ) if self.getResetPerPartition() else DataFrameUtils.zip_with_index(
+            df=the_df,
+            start_index=1,
+            col_name=self.getOutputCol()
+        )
+
+    @timefunc
+    def _fit(self, df: DataFrame) -> IdIndexerModel:
+        return IdIndexerModel(
+            self.input_col, self.partition_key, self.output_col, self._make_vocab_df(df).cache()
+        )
+
+
+class MultiIndexerModel(Transformer):
+    def __init__(self, models: List[IdIndexerModel]):
+        super().__init__()
+        self.models = models
+
+    def get_model_by_input_col(self, input_col):
+        for m in self.models:
+            if m.input_col == input_col:
+                return m
+
+        return None
+
+    def get_model_by_output_col(self, output_col):
+        for m in self.models:
+            if m.output_col == output_col:
+                return m
+
+        return None
+
+    @timefunc
+    def undo_transform(self, df: DataFrame) -> DataFrame:
+        curr_df = df.cache()
+
+        for model in self.models:
+            curr_df = model.undo_transform(curr_df).cache()
+
+        return curr_df
+
+    @timefunc
+    def _transform(self, df: DataFrame) -> DataFrame:
+        curr_df = df.cache()
+
+        for model in self.models:
+            curr_df = model.transform(curr_df).cache()
+
+        return curr_df
+
+
+class MultiIndexer(Estimator):
```

Comment:
This class seems like it duplicating the logic of a `Pipeline` might not be necessary in your code

Reply:
Multiindexer([indexer1, indexer2]) => Pipeline(stages=[indexer1, indexer2])

Reply:
I started to make this change but I had to roll it back as MultiIndexerModel (which is created by MultiIndexer) has the following methods which are necessary: 
* get_model_by_input_col
* get_model_by_output_col
* undo_transform
Hence MultiIndexerModel is not equivalent to PipelineModel hence MultiIndexer is also not equivalent to Pipeline